src/hnsw/build.c

```diff
@@ -153,6 +158,56 @@ void CheckHnswIndexDimensions(Relation index, Datum arrayDatum, int dimensions)
     }
 }
 
+/*
+ * Infer the dimensionality of the index from the heap
+ */
+static int GetArrayLengthFromHeap(Relation heap, IndexInfo *indexInfo, int currDim)
+{
+#if PG_VERSION_NUM < 120000
+    HeapScanDesc scan;
+#else
+    TableScanDesc scan;
+#endif
+    HeapTuple  tuple;
+    Snapshot   snapshot;
+    ArrayType *array;
+    Datum      datum;
+    bool       isNull;
+    int        n_items = currDim;
+    int        indexCol;
+
+    // If NumIndexAttrs isn't 1 there's no clear way to infer the dim
+    if(indexInfo->ii_NumIndexAttrs != 1) {
+        return n_items;
+    }
+
+    indexCol = indexInfo->ii_IndexAttrNumbers[ 0 ];
+
+    // Get the first row off the heap
+    snapshot = GetTransactionSnapshot();
+#if PG_VERSION_NUM < 120000
+    scan = heap_beginscan(heap, snapshot, 0, NULL);
+#else
+    scan = heap_beginscan(heap, snapshot, 0, NULL, NULL, SO_TYPE_SEQSCAN);
+#endif
+    tuple = heap_getnext(scan, ForwardScanDirection);
+    if(tuple == NULL) {
+        heap_endscan(scan);
+        return n_items;
+    }
+
+    // Get the indexed column out of the row and return it's dimensions
+    datum = heap_getattr(tuple, indexCol, RelationGetDescr(heap), &isNull);
+    if(!isNull) {
```

Comment:
So it seems, if the first row in the table has NULL vector and all the ones afterwards have e.g. 3 dimensional real[] arrays, then we fail to infer. This is fair. Let's just add a test for it and document it